ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -78,6 +80,50 @@ private[kvutils] object InputsAndEffects {
       parties.toList.sorted.map(partyStateKey)
     }
 
+    def addOrUpdateResolvedContractId(
+        key: DamlContractKey,
+        contractId: Option[Value.ContractId]): Unit =
+      resolvedContractIdsMap.get(key) match {
+        case None => resolvedContractIdsMap += (key -> contractId)
+        case _ => ()
+      }
+
+    def updateMappingWithExercisesNode(exe: Node.NodeExercises[NodeId, Value.ContractId]): Unit =
+      exe.key.foreach { keyWithMaintainers =>
+        val key =
+          globalKeyToStateKey(GlobalKey(exe.templateId, forceNoContractIds(keyWithMaintainers.key)))
+        addOrUpdateResolvedContractId(key.getContractKey, Some(exe.targetCoid))
+      }
+
+    def stateKey(
+        node: LeafOnlyNode[Value.ContractId],
+        keyWithMaintainers: KeyWithMaintainers[Value[Value.ContractId]]): DamlStateKey =
+      globalKeyToStateKey(GlobalKey(node.templateId, forceNoContractIds(keyWithMaintainers.key)))
+
+    def updateMappingWithLeafNode(node: LeafOnlyNode[Value.ContractId]): Unit = {
+      node match {
+        case fetch: Node.NodeFetch[Value.ContractId] =>
+          fetch.key.foreach { keyWithMaintainers =>
+            val key = stateKey(fetch, keyWithMaintainers)
+            addOrUpdateResolvedContractId(key.getContractKey, Some(fetch.coid))
+          }
+        case create: Node.NodeCreate[Value.ContractId] =>
+          create.key.foreach { keyWithMaintainers =>
+            val key = stateKey(create, keyWithMaintainers)
+            addOrUpdateResolvedContractId(key.getContractKey, Some(create.coid))
```

Comment:
```suggestion
            addOrUpdateResolvedContractId(key.getContractKey, None)
```
Before a contract is created, there shouldn't be a contract with the same key. So you want to check for `None` here.

Reply:
Thanks for this comment! It might make validation in `TransactionCommitter` cleaner too.

Reply:
Fixed.